server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java

```diff
@@ -198,6 +214,57 @@ final class DefaultSearchContext extends SearchContext {
         }
     }
 
+    static long getFieldCardinality(String field, IndexService indexService, DirectoryReader directoryReader) {
+        MappedFieldType mappedFieldType = indexService.mapperService().fieldType(field);
+        if (mappedFieldType != null) {
+            IndexFieldData<?> indexFieldData = indexService.loadFielddata(
+                mappedFieldType,
+                FieldDataContext.noRuntimeFields("field cardinality")
```

Comment:
I saw one of the test fail with: 

```
[2023-11-24T10:42:37,504][WARN ][r.suppressed             ] [node-0] path: /seats/_search, params: {index=seats, error_trace=true}, status: 500 Failed to execute phase [query], all shards failed; shardFailures {[EzGoEeSBSve92wrxPqIJlw][seats][0]: org.elasticsearch.transport.RemoteTransportException: [node-0][127.0.0.1:38479][indices:data/read/search[phase/query]]
--
  | »  Caused by: java.lang.UnsupportedOperationException: Runtime fields not supported for [field cardinality]
  | »       at org.elasticsearch.server@8.12.0-
```

Maybe we should check here the mappedFieldType? And return -1 if the type isn't a keyword field?

Reply:
I am going to catch exceptions on loading field data, that seems safer than another instanceof check. If it throws exception, we can't get the cardinality, hence we don't parallelize?

Reply:
I was also thinking about other use cases. Like terms agg on a number field. But I think catching exception should cover this. So +1